File: 09-Prototype/src/go/registry.go
package main

import "fmt"

type ShapeRegistry struct {
	prototypes map[string]Shape
}

func NewShapeRegistry() *ShapeRegistry {
	return &ShapeRegistry{prototypes: make(map[string]Shape)}
}

func (r *ShapeRegistry) Register(key string, shape Shape) {
	r.prototypes[key] = shape
}

func (r *ShapeRegistry) Get(key string) Shape {
	prototype, ok := r.prototypes[key]
	if !ok {
		panic(fmt.Sprintf("Unknown prototype: %s", key))
	}
	return prototype.Clone() // return a CLONE, not the original
}

func (r *ShapeRegistry) Contains(key string) bool {
	_, ok := r.prototypes[key]
	return ok
}

func (r *ShapeRegistry) Size() int {
	return len(r.prototypes)
}

File: 09-Prototype/src/rust/registry.rs
use std::collections::HashMap;
use crate::shapes::Prototype;

pub struct ShapeRegistry {
    prototypes: HashMap<String, Box<dyn Prototype>>,
}

impl ShapeRegistry {
    pub fn new() -> Self {
        ShapeRegistry { prototypes: HashMap::new() }
    }

    pub fn register(&mut self, key: &str, shape: Box<dyn Prototype>) {
        self.prototypes.insert(key.to_string(), shape);
    }

    pub fn get(&self, key: &str) -> Box<dyn Prototype> {
        let prototype = self.prototypes.get(key)
            .unwrap_or_else(|| panic!("Unknown prototype: {}", key));
        prototype.clone_shape()  // return a CLONE, not the original
    }

    pub fn contains(&self, key: &str) -> bool {
        self.prototypes.contains_key(key)
    }

    pub fn size(&self) -> usize {
        self.prototypes.len()
    }
}
